File: go/kan/invar_day.go
package capitan

type Invar_Day[F, G, A any] struct {
	Fb  F
	Gc  G
	Bca func(b, c any) A
	Abc func(a A) (b any, c any)
}

func day[F, G, A, B, C any](fa F, gb G) Invar_Day[F, G, A] {
	return Invar_Day[F, G, A]{
		Fb: fa,
		Gc: gb,
		Bca: func(b B, c C) A {
			return A(B(b), C(c))
		},
		Abc: func(a A) (B, C) {
			b, c := interface{}(a).(A)
			return B(b), C(c)
		},
	}
}

func invmap[F, G, A, B, C, D any](f func(A) B, g func(C) D, day Invar_Day[F, G, A]) Invar_Day[F, G, B] {
	return Invar_Day[F, G, B]{
		Fb: day.Fb,
		Gc: day.Gc,
		Bca: func(b B, c D) B {
			return f(day.Bca(interface{}(b).(A), interface{}(c).(C)))
		},
		Abc: func(a B) (B, D) {
			b, c := day.Abc(interface{}(a).(A))
			return B(b), D(g(interface{}(c).(C)))
		},
	}
}

func assoc[F, G, H, A, B, C any](day Invar_Day[F, Invar_Day[G, H, C], A]) Invar_Day[Invar_Day[F, G, C], H, A] {
	return Invar_Day[Invar_Day[F, G, C], H, A]{
		Fb: day.Fb,
		Gc: day.Gc.Gc,
		Bca: func(bc Invar_Day[F, G, C], e H) A {
			return day.Bca(bc.Fb, day.Gc.Bca(bc.Gc, e))
		},
		Abc: func(a A) (Invar_Day[F, G, C], H) {
			b, c := day.Abc(a)
			d, e := day.Gc.Abc(c)
			return Invar_Day[F, G, C]{day.Fb, d, day.Bca, func(a A) (B, C) { return b, interface{}(a).(C) }}, e
		},
	}
}

func disassoc[F, G, H, A, B, C any](day Invar_Day[Invar_Day[F, G, B], H, A]) Invar_Day[F, Invar_Day[G, H, A], B] {
	return Invar_Day[F, Invar_Day[G, H, A], B]{
		Fb: day.Fb.Fb,
		Gc: Invar_Day[G, H, A]{day.Fb.Gc, day.Gc, day.Fb.Bca, day.Abc},
		Bca: func(b B, day Invar_Day[G, H, A]) B {
			return day.Bca(day.Fb.Bca(b, day.Gc), day.Abc)
		},
		Abc: func(b B) (B, Invar_Day[G, H, A]) {
			a, c := day.Abc(day.Bca(day.Fb.Abc(b), day.Gc))
			return b, Invar_Day[G, H, A]{day.Fb.Gc, c, day.Fb.Bca, a}
		},
	}
}

func swapped[F, G, A any](day Invar_Day[F, G, A]) Invar_Day[G, F, A] {
	return Invar_Day[G, F, A]{
		Fb: day.Gc,
		Gc: day.Fb,
		Bca: func(c G, b F) A {
			return day.Bca(interface{}(b).(A), interface{}(c).(C))
		},
		Abc: func(a A) (G, F) {
			b, c := day.Abc(a)
			return interface{}(c).(G), interface{}(b).(F)
		},
	}
}

func intro1[F, A any](fa F) Invar_Day[unit, F, A] {
	return Invar_Day[unit, F, A]{
		Fb:  unit{},
		Gc:  fa,
		Bca: func(_ unit, a A) A {
	    return a
		},
		Abc: func(a A) (unit, A) {
		    return unit{}, a
		},
	}
}

func intro2[F, A any](fa F) Invar_Day[F, unit, A] {
    return Invar_Day[F, unit, A]{
        Fb: fa,
        Gc: unit{},
        Bca: func(a A, _ unit) A {
            return a
        },
        Abc: func(a A) (A, unit) {
            return a, unit{}
        },
    }
}

func elim1[F, A any](day Invar_Day[unit, F, A]) F {
    return invmap(day.Bca(unit{}), func(a A) A { return a }, day.Gc)
}

func elim2[F, A any](day Invar_Day[F, unit, A]) F {
    return invmap(func(a A) A { return day.Bca(a, unit{}) }, func(_ unit) unit { return unit{} }, day.Fb)
}

func trans1[F, G, H, A, B any](fg func(F) G, day Invar_Day[F, H, A]) Invar_Day[G, H, A] {
    return Invar_Day[G, H, A]{
        Fb: fg(day.Fb),
        Gc: day.Gc,
        Bca: day.Bca,
        Abc: day.Abc,
    }
}

func trans2[F, G, H, A, B any](gh func(G) H, day Invar_Day[F, G, A]) Invar_Day[F, H, A] {
    return Invar_Day[F, H, A]{
        Fb: day.Fb,
        Gc: gh(day.Gc),
        Bca: day.Bca,
        Abc: day.Abc,
    }
}

type Contravariant[F, G, A any] struct {
    Fb  F
    Gc  G
    Abc func(a A) (b any, c any)
}

func toContravariant[F, G, A any](day Invar_Day[F, G, A]) Contravariant[F, G, A] {
    return Contravariant[F, G, A]{
        Fb:  day.Fb,
        Gc:  day.Gc,
        Abc: day.Abc,
    }
}

type Covariant[F, G, A any] struct {
    Fb  F
    Gc  G
    Bca func(b any, c any) A
}

func toCovariant[F, G, A any](day Invar_Day[F, G, A]) Covariant[F, G, A] {
    return Covariant[F, G, A]{
        Fb:  day.Fb,
        Gc:  day.Gc,
        Bca: day.Bca,
    }
}

File: python/kan/invar_day.py
from typing import Any, Callable, Generic, Tuple, TypeVar

A = TypeVar('A')
B = TypeVar('B')
C = TypeVar('C')
D = TypeVar('D')
E = TypeVar('E')
F = TypeVar('F')
G = TypeVar('G')
H = TypeVar('H')

class Invar_Day(Generic[F, G, A]):
    def __init__(self, fb: F, gc: G, bca: Callable[[Any, Any], A], abc: Callable[[A], Tuple[Any, Any]]):
        self.fb = fb
        self.gc = gc
        self.bca = bca
        self.abc = abc

    @staticmethod
    def day(fa: F, gb: G) -> 'Day[F, G, Tuple[B, C]]':
        def bca(b: Any, c: Any) -> Tuple[B, C]:
            return b, c
        def abc(a: Tuple[B, C]) -> Tuple[Any, Any]:
            return a[0], a[1]
        return Day(fa, gb, bca, abc)

    def invmap(self, f: Callable[[A], B], g: Callable[[C], D]) -> 'Day[F, G, B]':
        def bca(b: Any, c: Any) -> B:
            return f(self.bca(b, g(c)))
        def abc(a: B) -> Tuple[Any, Any]:
            b, c = self.abc(a)
            return b, g(c)
        return Day(self.fb, self.gc, bca, abc)

    def assoc(self) -> 'Day[Day[F, G, D], H, A]':
        def f(a: A) -> Tuple['Day[F, G, D]', H]:
            b, c = self.abc(a)
            d, e = self.gc.abc(c)
            return Day(self.fb, d, lambda b, c: self.bca(b, c), lambda a: (b, a)), e
        def g(bc: 'Day[F, G, D]', e: H) -> A:
            return self.bca(bc.fb, self.gc.bca(bc.gc, e))
        return Day(Day(self.fb, self.gc.gc, lambda b, c: b, lambda c: (c, ())), self.gc.gc, g, f)

    def disassoc(self) -> 'Day[F, Day[G, H, A], B]':
        def f(b: B, day: 'Day[G, H, A]') -> B:
            return day.bca(self.fb.bca(b, day.gc), day.abc)
        def g(b: B) -> Tuple[B, 'Day[G, H, A]']:
            a, c = self.abc(self.bca(self.fb.abc(b), self.gc))
            return b, Day(self.fb.gc, c, self.fb.bca, a)
        return Day(self.fb.fb, Day(self.fb.gc, self.gc, self.fb.bca, self.abc), f, g)

    def swapped(self) -> 'Day[G, F, A]':
        def bca(c: G, b: F) -> A:
            return self.bca(b, c)
        def abc(a: A) -> Tuple[G, F]:
            b, c = self.abc(a)
            return c, b
        return Day(self.gc, self.fb, bca, abc)

    def intro1(self, fa: F) -> 'Day[tuple, F, B]':
        def bca(_, a: B) -> B:
            return a
        def abc(a: B) -> Tuple[tuple, B]:
            return (), a
        return Day((), fa, bca, abc)

    def intro2(self, fa: F) -> 'Day[F, tuple, B]':
        def bca(a: B, _) -> B:
            return a
        def abc(a: B) -> Tuple[B, tuple]:
            return a, ()
        return Day(fa, (), bca, abc)

    def elim1(self) -> F:
        return self.invmap(lambda a: self.bca((), a), lambda _:()).gc

    def elim2(self) -> F:
        return self.invmap(lambda a: self.bca(a, ()), lambda _: ()).fb

    def trans1(self, fg: Callable[[F], G]) -> 'Day[G, H, A]':
        return Day(fg(self.fb), self.gc, self.bca, self.abc)

    def trans2(self, gh: Callable[[G], H]) -> 'Day[F, H, A]':
        return Day(self.fb, gh(self.gc), self.bca, self.abc)

    def to_contravariant(self):
        return Contravariant(self.fb, self.gc, self.abc)

    def to_covariant(self):
        return Covariant(self.fb, self.gc, self.bca)

class Contravariant(Generic[F, G, A]):
    def __init__(self, fb: F, gc: G, abc: Callable[[A], Tuple[Any, Any]]):
        self.fb = fb
        self.gc = gc
        self.abc = abc

class Covariant(Generic[F, G, A]):
    def __init__(self, fb: F, gc: G, bca: Callable[[Any, Any], A]):
        self.fb = fb
        self.gc = gc
        self.bca = bca